src/VisualStudio/IntegrationTest/IntegrationTests/AbstractIntegrationTest.cs

```diff
@@ -5,26 +5,38 @@
 using Microsoft.VisualStudio.IntegrationTest.Utilities.OutOfProcess;
 using System;
 using System.Threading;
+using System.Threading.Tasks;
 using System.Windows.Automation;
+using Xunit;
 
 namespace Roslyn.VisualStudio.IntegrationTests
 {
     [CaptureTestName]
-    public abstract class AbstractIntegrationTest : IDisposable
+    public abstract class AbstractIntegrationTest : IAsyncLifetime, IDisposable
     {
-        public readonly VisualStudioInstance VisualStudio;
-
         protected readonly string ProjectName = "TestProj";
         protected readonly string SolutionName = "TestSolution";
 
+        private readonly VisualStudioInstanceFactory _instanceFactory;
         private VisualStudioInstanceContext _visualStudioContext;
 
-        protected AbstractIntegrationTest(
-            VisualStudioInstanceFactory instanceFactory)
+        protected AbstractIntegrationTest(VisualStudioInstanceFactory instanceFactory)
         {
+            Assert.Equal(ApartmentState.STA, Thread.CurrentThread.GetApartmentState());
+            _instanceFactory = instanceFactory;
             Automation.TransactionTimeout = 20000;
-            _visualStudioContext = instanceFactory.GetNewOrUsedInstance(SharedIntegrationHostFixture.RequiredPackageIds);
-            VisualStudio = _visualStudioContext.Instance;
+        }
+
+        public VisualStudioInstance VisualStudio => _visualStudioContext?.Instance;
+
+        public virtual async Task InitializeAsync()
+        {
+            _visualStudioContext = await _instanceFactory.GetNewOrUsedInstanceAsync(SharedIntegrationHostFixture.RequiredPackageIds).ConfigureAwait(false);
+        }
+
+        public Task DisposeAsync()
+        {
+            return Task.CompletedTask;
```

Comment:
This seems rather weird. We're initializing via IAsyncLifetime, but we don't dispose via IAsyncLifetime? This feels like it's a bug.

Reply:
It's possible this could be changed, but keep in mind it's just doing the same thing for cleanup that we're doing for initialization:

Constructor &rarr; `InitializeAsync` &rarr; *Test* &rarr; `DisposeAsync` &rarr; `Dispose`

I needed to move to `InitializeAsync` to fix a deadlock where `Result` was accessed instead of using `await`. I didn't need to make any changes to `Dispose` for things to work.

Reply:
I agree this pattern looks very strange and is difficult to reason about. Can we at least change the method signature so it is an explicit implementation?

Reply:
:memo: On further discussion with @mavasani, I decided to keep the current form because the non-virtual method make it more difficult to create a shadowing method in a derived type without being forced to reason about all of this. If I made it explicitly implement the interface method, it would be easy to end up with a hierarchy where `DisposeAsync` was called at different levels unequally.